test/examples/ucp_client_server.c

```diff
@@ -0,0 +1,436 @@
+                           socklen_t addrlen, ucp_listener_h *listener,
+                           ucx_server_ctx_t *context)
+{
+    ucp_listener_params_t params;
+    ucs_status_t status;
+
+    params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    params.sockaddr.addr      = addr;
+    params.sockaddr.addrlen   = addrlen;
+    params.accept_handler.cb  = server_accept_cb;
+    params.accept_handler.arg = context;
+
+    status = ucp_listener_create(ucp_worker, &params, listener);
+    if (status == UCS_OK) {
+        printf("Waiting for connection...\n");
+    } else {
+        fprintf(stderr, "failed to listen\n");
+    }
+
+    return status;
+}
+
+/**
+ * Set an address for the server to listen on - INADDR_ANY on a well known port.
+ */
+void set_listen_addr(struct sockaddr_in *listen_addr)
+{
+    /* The server will listen on INADDR_ANY */
+    memset(listen_addr, 0, sizeof(struct sockaddr_in));
+    listen_addr->sin_family      = AF_INET;
+    listen_addr->sin_addr.s_addr = INADDR_ANY;
+    listen_addr->sin_port        = server_port;
+}
+
+/**
+ * Set and address to connect to. A given IP address on a well known port.
+ */
+void set_connect_addr(const char *ip, struct sockaddr_in *connect_addr)
+{
+    memset(connect_addr, 0, sizeof(struct sockaddr_in));
+    connect_addr->sin_family      = AF_INET;
+    connect_addr->sin_addr.s_addr = inet_addr(ip);
+    connect_addr->sin_port        = server_port;
+}
+
+/**
+ * Initialize the server side. The server starts listening on the set address
+ * and waits for its connected endpoint to be created.
+ */
+static int start_server(ucp_worker_h ucp_worker, ucx_server_ctx_t *context,
+                        ucp_listener_h *listener)
+{
+    struct sockaddr_in listen_addr;
+    ucs_status_t status;
+
+    context->ep = NULL;
+
+    set_listen_addr(&listen_addr);
+    status = server_listen(ucp_worker, (const struct sockaddr*)&listen_addr,
+                           sizeof(listen_addr), listener,
+                           context);
+
+    if (status == UCS_OK) {
+        while (context->ep == NULL) {
+            ucp_worker_progress(ucp_worker);
+        }
+    }
+
+    return status;
+}
+
+/**
+ * Initialize the client side. Create an endpoint from the client side to be
+ * connected to the remote server (to the given IP).
+ */
+static ucs_status_t start_client(ucp_worker_h ucp_worker, const char *ip,
+                                 ucp_ep_h *client_ep)
+{
+    ucp_ep_params_t ep_params;
+    struct sockaddr_in connect_addr;
+    ucs_status_t status;
+
+    set_connect_addr(ip, &connect_addr);
+
+    ep_params.field_mask       = UCP_EP_PARAM_FIELD_FLAGS     |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
```

Comment:
Is this really needed for this example? If you are going to handle errors, how about to define error handling callback too?

Reply:
@evgeny-leksikov since in this case i'm using UCP_EP_CLOSE_MODE_FORCE, seems that this err_mode is required:
https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_ep.c#L601
AFAIU this code example should be basic for the client-server API, so probably shouldn't add additional features here. I plan to remove this though when we can switch to FLUSH.
btw, error handling is tested with the client-server api in gtest - both for uct and ucp.

Reply:
IC, but first of all this is not a test, this is an example. Every line should be clear for user. Pls, add error handler callback how to handle possible error or comment why you set error handling mode to peer mode and ignore its handling by lack of CB.